src/client.ts

```diff
@@ -343,23 +353,46 @@ class OptimizelyReactSDKClient implements ReactSDKClient {
   }
 
   public async fetchQualifiedSegments(options?: optimizely.OptimizelySegmentOption[]): Promise<boolean> {
-    if (!this.userContext || !this.isReady()) {
+    if (this.odpExplicitlyOff) {
+      return true;
+    }
+
+    if (!this.userContext) {
       return false;
     }
 
     return await this.userContext.fetchQualifiedSegments(options);
   }
 
   public async setUser(userInfo: UserInfo): Promise<void> {
-    this.setCurrentUserContext(userInfo);
-
     this.user = {
       id: userInfo.id || DefaultUser.id,
       attributes: userInfo.attributes || DefaultUser.attributes,
     };
 
-    if (this.getIsReadyPromiseFulfilled()) {
-      await this.fetchQualifiedSegments();
+    let fetchQualifedSegmentsSucceed = false;
+    if (userInfo.id === DefaultUser.id) {
+      // if user is anonymous...
+      // wait for the SDK client to be ready before
+      await this._client?.onReady();
+      // setting the user context
+      this.setCurrentUserContext(userInfo);
+
+      this.user.id = this.userContext?.getUserId() || DefaultUser.id;
+
+      fetchQualifedSegmentsSucceed = await this.fetchQualifiedSegments();
+    } else {
+      this.setCurrentUserContext(userInfo);
+
+      this.user.id = this.userContext?.getUserId() || DefaultUser.id;
+
+      await this._client?.onReady();
+      fetchQualifedSegmentsSucceed = await this.fetchQualifiedSegments();
```

Comment:
This and the block above for "DefaultUser.id" look same.
@raju-opti can you clarify your concern about this having same async for both?

Reply:
I agree. These are very close or could be DRY-ed out.  Whether to set the underlying userContext or try to fetchQualifiedSegments, we still need to await the JS SDK's ready.

Waiting a little bit for @raju-opti but we should move forward when ready.